nvtabular/loader/tensorflow.py

```diff
@@ -416,8 +422,30 @@ def _get_sparse_tensor(self, values, indices, num_rows, seq_limit):
         return sparse_tensor
 
     def _build_sparse_tensor(self, values, offsets, diff_offsets, num_rows, seq_limit):
+        """Builds sparse tensors in the TensorFlow dataloader.
+
+        Parameters
+        ----------
+        values :
+        offsets :
+        diff_offsets :
+        num_rows :
+        seq_limit :
+
+        Returns
+        -------
+        tf.sparse
+            Our built TensorFlow sparse tensor.
+
+        """
         ragged = tf.RaggedTensor.from_row_lengths(values=values, row_lengths=diff_offsets)
-        tensor = tf.RaggedTensor.from_tensor(ragged.to_tensor(shape=[None, seq_limit])).to_sparse()
+        if self.pad_left:
+            max_len = max(max(len(row) for row in ragged), seq_limit)
+            tensor = tf.stack([tf.pad(row, [[max_len - len(row), 0]]) for row in ragged], axis=0)
```

Comment:
Is iterating through each row the only way to do this? You can actually logically figure out all the padding amounts by doing array math and then you can pass that entire list of "pad_length" entries at once that way its not doing each row, one at a time.

Reply:
This is good to know. Let me see how to implement this approach you outline here.

Reply:
I am having a hard time seeing currently a vectorized or further optimized approach to this code snippet. Figuring out the padding amounts by row is not that difficult, but what to do with that tensor of row lengths is very much unclear to me. For instance there is `tf.pad()`, but this only takes padding tensors of shape `[n, 2]`, where `n` is the rank of the original tensor, so that we may only do a constant amount of padding per dimension. For `torch` there is `torch.nn.functional.pad()`, which also only does a constant amount of left or right padding per dimension. There is `tf.ragged`'s `to_tensor()` method, which implicitly does padding or truncation according to a given shape, but this only does padding on the right. So it is not that clear how to do a further optimized approach using this variable padding tensor. Did you have some other `tensorflow` or `torch` methods that you had in mind, or some other vectorized approach using the variable padding lengths that I am missing? I have tried some others, but to no success yet. I can modify the run-length-encoding of the ragged tensor here, but I would guess that this would also not be a vectorized operation.

Reply:
@jperez999 How does this current updated implementation of this TensorFlow code snippet without iterating through the rows of the ragged tensor address your feedback?

Reply:
The for loops you are doing are the problem here I think... That methodology is extremely slow. You have to go in to each row read the data... as opposed to doing the entire column at once.  
```
digits = tf.ragged.constant([[3, 1, 4, 1], [], [5, 9, 2], [6], []])
padded = tf.reverse(digits, [-1]).to_tensor(0)
final_tensor = tf.reverse(padded, [-1])
```
I think this gives you the padding behavior you want and its a tad bit faster I am clocking this logic in at 0.005383729934692383 and I am clocking your logic at 0.026534557342529297.

Reply:
@jperez999 Did you see the updated TensorFlow implementation that I have here that does not use `for` loops? It is at commit 01749f9.

Reply:
that commit you linked there references conda environment files... no code changes... As the PR stands that for loop logic still exists for both torch and tensorflow.

Reply:
That commit is a final merge of main into this branch. The changes are in the predecessors of this commit. They should be viewable [here](https://github.com/NVIDIA/NVTabular/pull/1126/files) in the web UI or by doing a checkout of that commit. The `for` loop implementation for TensforFlow is an outdated change. I also tagged you with a comment on the code section below. Are you able to see these changes now?

Reply:
In particular, in my web UI, the `tensorflow` code snippet you are referencing has a yellow "outdated" box next to it.